scripts/egp27/createProposalUpdateGrants.ts

```diff
@@ -0,0 +1,236 @@
+  signer: Signer,
+  provider: Provider,
+  coreVotingAddress: string,
+  timeLockAddress: string,
+  // vesting vault proxy address
+  vestingVaultAddress: string,
+  // upgraded vesting vault implementation address
+  unfrozenVaultAddress: string,
+  // original vesting vault implementation address
+  frozenVaultAddress: string,
+  // voting vaults to query vote power from to submit proposal
+  votingVaultAddresses: string[],
+  // extra data for voting vaults if necessary
+  extraVaultData: string[],
+  // grants to add/remove
+  grants: Grant[]
+): Promise<ProposalInfo> {
+  /********************************************************************************
+   * Set up a new proposal.  This proposal will perform 3 actions:
+   *   1. temporarily upgrade the VestingVault's implementation to one that can add AND remove grants
+   *   2. perform necessary updates to grants
+   *   3. reset the VestingVault back to the original implementation.
+   ********************************************************************************/
+
+  // step 1 is to update the vesting vault implementation address
+  const {
+    targets,
+    callDatas,
+    proposalHash,
+    targetsTimeLock,
+    calldatasTimeLock,
+    callHashTimelock,
+  } = await getUpdateGrantsProposalArgs(
+    provider,
+    grants,
+    unfrozenVaultAddress,
+    frozenVaultAddress,
+    vestingVaultAddress,
+    timeLockAddress
+  );
+
+  const ballot = 0; // 0 - YES, 1 - NO, 2 - ABSTAIN
+
+  // create the arguments to coreVoting.proposal()
+  const coreVotingContract = CoreVoting__factory.connect(
+    coreVotingAddress,
+    signer
+  );
+
+  // last chance to execute to vote is ~14 days from current block
+  const currentBlock = await provider.getBlockNumber();
+  const lastCall = Math.round(DAY_IN_BLOCKS * 14 + currentBlock);
+
+  const tx = await coreVotingContract.proposal(
+    votingVaultAddresses,
+    extraVaultData,
+    targets,
+    callDatas,
+    lastCall,
+    ballot
+  );
+  await tx.wait(1);
+
+  // just getting the proposalId
+  const proposalCreatedEvents = await coreVotingContract.queryFilter(
+    coreVotingContract.filters.ProposalCreated(),
+    currentBlock
+  );
+  const proposalId = proposalCreatedEvents[0].args[0].toNumber();
+
+  const proposalArgs = [
+    ["proposalId", proposalId],
+    ["votingVaults", votingVaultAddresses],
+    ["extraVaultData", extraVaultData],
+    ["targets", targets],
+    ["callDatas", callDatas],
+    ["proposalHash", proposalHash],
+    ["targetsTimeLock", targetsTimeLock],
+    ["calldatasTimeLock", calldatasTimeLock],
+    ["callHashTimelock", callHashTimelock],
+    ["lastCall", lastCall],
+    ["ballot", ballot],
+  ];
+
+  console.log("Proposal created with:");
+  proposalArgs.forEach(([name, value]) => console.log(name, value));
+
+  const proposalInfo: ProposalInfo = Object.fromEntries(proposalArgs);
+  return proposalInfo;
+}
+
+export async function getUpdateGrantsProposalArgs(
```

Comment:
Just to make sure I understand the way that proposals work, it seems that we are:

1. Construct several function calls to the vesting vault proxy. This will (1) update the contract to the unfrozen vesting vault, (2) add/remove a bunch of grants, and (3) reset the vesting vault back to the frozen vesting vault.
2. We take the hash of the targets and calldata and use this as an argument to `registerCall`. This `registerCall` is the primary action of the proposal and it starts the clock for the real actions to be executed against the time lock.

Let me know if I'm missing anything.

Reply:
Yeah you pretty much hit the nail on the head here.  The actual proposal in CoreVoting is like you said:
 - a target of the Timelock
 -  a calldata of a `registerCall(timeLockCallHash)`, 
 
 Important to note that the proposalId is the hash of these two things above.
 
 The timeLockCallHash is the hash of the targets and calldatas you outlined in 1., the 'real' actions executed from the timelock.
 
 Also important to note that for anyone to verify the hash, they obviously need the 'real' targets and calldatas.